### PR TITLE
Swap fixes

### PR DESCRIPTION
**Pull request by FrederikBolding**
- Use `estimateGas` for trade tx
- Fix some issues with newer versions of Ethers
- Fix UX issues when not enough ETH
- Fix expiration countdown issue
- Move some StoreProvider logic to saga
- Make it easier to support more networks in the future 😉

### CODE REVIEW

**Comment by tayvano** [comment]
1. im getting a white screen onload with some other mycryptobuilds -- either #3835 or #3484

```
TypeError: Cannot read property 'name' of undefined
    at main.87a1488352a5fb861c66.js:300215
    at Array.map (<anonymous>)
    at G (main.87a1488352a5fb861c66.js:300215)
    at t.default (main.87a1488352a5fb861c66.js:300215)
    at lo (main.87a1488352a5fb861c66.js:209737)
    at Ss (main.87a1488352a5fb861c66.js:209737)
    at mu (main.87a1488352a5fb861c66.js:209737)
    at pu (main.87a1488352a5fb861c66.js:209737)
    at au (main.87a1488352a5fb861c66.js:209737)
    at main.87a1488352a5fb861c66.js:209737
fs @ main.87a1488352a5fb861c66.js:209737
Os.a.componentDidCatch.n.callback @ main.87a1488352a5fb861c66.js:209737
ga @ main.87a1488352a5fb861c66.js:209737
ba @ main.87a1488352a5fb861c66.js:209737
gu @ main.87a1488352a5fb861c66.js:209737
t.unstable_runWithPriority @ main.87a1488352a5fb861c66.js:210031
Vi @ main.87a1488352a5fb861c66.js:209737
bu @ main.87a1488352a5fb861c66.js:209737
au @ main.87a1488352a5fb861c66.js:209737
(anonymous) @ main.87a1488352a5fb861c66.js:209737
t.unstable_runWithPriority @ main.87a1488352a5fb861c66.js:210031
Vi @ main.87a1488352a5fb861c66.js:209737
qi @ main.87a1488352a5fb861c66.js:209737
Ki @ main.87a1488352a5fb861c66.js:209737
le @ main.87a1488352a5fb861c66.js:209737
In @ main.87a1488352a5fb861c66.js:209737
kn @ main.87a1488352a5fb861c66.js:209737
xn @ main.87a1488352a5fb861c66.js:209737
t.unstable_runWithPriority @ main.87a1488352a5fb861c66.js:210031
Cn @ main.87a1488352a5fb861c66.js:209737
02:15:20.707 main.87a1488352a5fb861c66.js:320334 TypeError: Cannot read property 'name' of undefined
    at main.87a1488352a5fb861c66.js:300215
    at Array.map (<anonymous>)
    at G (main.87a1488352a5fb861c66.js:300215)
    at t.default (main.87a1488352a5fb861c66.js:300215)
    at lo (main.87a1488352a5fb861c66.js:209737)
    at Ss (main.87a1488352a5fb861c66.js:209737)
    at mu (main.87a1488352a5fb861c66.js:209737)
    at pu (main.87a1488352a5fb861c66.js:209737)
    at au (main.87a1488352a5fb861c66.js:209737)
    at main.87a1488352a5fb861c66.js:209737
componentDidCatch @ main.87a1488352a5fb861c66.js:320334
Os.a.componentDidCatch.n.callback @ main.87a1488352a5fb861c66.js:209737
ga @ main.87a1488352a5fb861c66.js:209737
ba @ main.87a1488352a5fb861c66.js:209737
gu @ main.87a1488352a5fb861c66.js:209737
t.unstable_runWithPriority @ main.87a1488352a5fb861c66.js:210031
Vi @ main.87a1488352a5fb861c66.js:209737
bu @ main.87a1488352a5fb861c66.js:209737
au @ main.87a1488352a5fb861c66.js:209737
(anonymous) @ main.87a1488352a5fb861c66.js:209737
t.unstable_runWithPriority @ main.87a1488352a5fb861c66.js:210031
Vi @ main.87a1488352a5fb861c66.js:209737
qi @ main.87a1488352a5fb861c66.js:209737
Ki @ main.87a1488352a5fb861c66.js:209737
le @ main.87a1488352a5fb861c66.js:209737
In @ main.87a1488352a5fb861c66.js:209737
kn @ main.87a1488352a5fb861c66.js:209737
xn @ main.87a1488352a5fb861c66.js:209737
t.unstable_runWithPriority @ main.87a1488352a5fb861c66.js:210031
Cn @ main.87a1488352a5fb861c66.js:209737
02:15:21.039 main.87a1488352a5fb861c66.js:209737 TypeError: Cannot read property 'name' of undefined
    at main.87a1488352a5fb861c66.js:300215
    at Array.map (<anonymous>)
    at G (main.87a1488352a5fb861c66.js:300215)
    at t.default (main.87a1488352a5fb861c66.js:300215)
    at lo (main.87a1488352a5fb861c66.js:209737)
    at Ss (main.87a1488352a5fb861c66.js:209737)
    at mu (main.87a1488352a5fb861c66.js:209737)
    at pu (main.87a1488352a5fb861c66.js:209737)
    at au (main.87a1488352a5fb861c66.js:209737)
    at main.87a1488352a5fb861c66.js:209737
fs @ main.87a1488352a5fb861c66.js:209737
Ts.n.callback @ main.87a1488352a5fb861c66.js:209737
ga @ main.87a1488352a5fb861c66.js:209737
ba @ main.87a1488352a5fb861c66.js:209737
gu @ main.87a1488352a5fb861c66.js:209737
t.unstable_runWithPriority @ main.87a1488352a5fb861c66.js:210031
Vi @ main.87a1488352a5fb861c66.js:209737
bu @ main.87a1488352a5fb861c66.js:209737
au @ main.87a1488352a5fb861c66.js:209737
(anonymous) @ main.87a1488352a5fb861c66.js:209737
t.unstable_runWithPriority @ main.87a1488352a5fb861c66.js:210031
Vi @ main.87a1488352a5fb861c66.js:209737
qi @ main.87a1488352a5fb861c66.js:209737
Ki @ main.87a1488352a5fb861c66.js:209737
le @ main.87a1488352a5fb861c66.js:209737
In @ main.87a1488352a5fb861c66.js:209737
kn @ main.87a1488352a5fb861c66.js:209737
xn @ main.87a1488352a5fb861c66.js:209737
t.unstable_runWithPriority @ main.87a1488352a5fb861c66.js:210031
Cn @ main.87a1488352a5fb861c66.js:209737
02:15:21.040 main.87a1488352a5fb861c66.js:300215 Uncaught TypeError: Cannot read property 'name' of undefined
    at main.87a1488352a5fb861c66.js:300215
    at Array.map (<anonymous>)
    at G (main.87a1488352a5fb861c66.js:300215)
    at t.default (main.87a1488352a5fb861c66.js:300215)
    at lo (main.87a1488352a5fb861c66.js:209737)
    at Ss (main.87a1488352a5fb861c66.js:209737)
    at mu (main.87a1488352a5fb861c66.js:209737)
    at pu (main.87a1488352a5fb861c66.js:209737)
    at au (main.87a1488352a5fb861c66.js:209737)
    at main.87a1488352a5fb861c66.js:209737
```

2. if user has no tokens (to swap), scan for tokens. (feel free to convert this a card for down the road, this is a nice to have but not necessary)

**Comment by tayvano** [comment]
Ledger thinks I denied the transaction even though I didnt and it didnt even seem like it ried?
S 2021-05-13 at 5 19 16 AM

**Comment by tayvano** [comment]
jk lol it was becuase i had reset my ledger and the `contract data` setting was set to `NOT Allowed`. The desktop app showed `Ledger device: Invalid data received (0x6a80)` which is what clued me in.

**Review by Mrtenz — DISMISSED** [review]
Code looks fine. Should be tested before merging.

**Comment by FrederikBolding** [comment]
Potentially needs a small copy update etc.

File: src/features/SwapAssets/components/SwapAssets.tsx [hunk 1]
```diff
@@ -89,10 +91,12 @@ const SwapAssets = (props: Props) => {
     isEstimatingGas,
     expiration,
     isDemoMode,
-    baseAsset,
     settings
   } = props;
 
+  const network = useSelector(selectNetwork(selectedNetwork)) as Network;
+  const baseAsset = useSelector(getBaseAssetByNetwork(network)) as ExtendedAsset;
```
Comment (Mrtenz) [inline src/features/SwapAssets/components/SwapAssets.tsx:98]: Why these type casts? It should be able to infer the types from the selector?

File: src/features/SwapAssets/components/SwapAssets.tsx [hunk 2]
```diff
@@ -321,11 +324,8 @@ const SwapAssets = (props: Props) => {
 };
 
 const mapStateToProps = (state: AppState) => {
-  const network = selectDefaultNetwork(state) as Network;
-
   return {
     isDemoMode: getIsDemoMode(state),
-    baseAsset: getBaseAssetByNetwork(network)(state) as ExtendedAsset,
     settings: getSettings(state)
   };
```
Comment (Mrtenz) [inline src/features/SwapAssets/components/SwapAssets.tsx:330]: Why does some stuff use `mapStateToProps`, and other stuff `useSelector`? Maybe just move it all to hooks.

**Review by Mrtenz — CHANGES_REQUESTED** [review]

File: src/features/SwapAssets/components/SwapAssets.tsx (hunk 1 above)
Comment (FrederikBolding) [reply src/features/SwapAssets/components/SwapAssets.tsx:98]: It is not able to infer it when using functions that create selectors it seems.

**Review by Mrtenz — APPROVED** [review]

**Pull request merged**